### PR TITLE
Fix for detector `UncallableMethodOfAnonymousClass` to not report unused methods of method-local enumerations and records

### PR DESCRIPTION
**Pull request by baloghadamsoftware**
During compilation, enumerations and records are turned into Java classes with some auto-generated methods. Some of these methods are called by the code but some of them not. However, this is not a programming error or bad practice, these methods must not be considered as dead code because their generation cannot be prevented. Therefore no bug report `UMAC_UNCALLABLE_METHOD_OF_ANONYMOUS_CLASS` must be issued to such classes even if they are defined inside a method. See issue ([#2120](https://github.com/spotbugs/spotbugs/issues/2120))

----

Make sure these boxes are checked before submitting your PR -- thank you!

- [X] Added an entry into `CHANGELOG.md` if you have changed SpotBugs code

### CODE REVIEW

File: spotbugs/src/main/java/edu/umd/cs/findbugs/detect/UncallableMethodOfAnonymousClass.java [hunk 1]
```diff
@@ -59,6 +59,11 @@ public UncallableMethodOfAnonymousClass(BugReporter bugReporter) {
 
     @Override
     public void visitJavaClass(JavaClass obj) {
+        if (Values.DOTTED_JAVA_LANG_ENUM.equals(obj.getSuperclassName()) ||
+                Values.DOTTED_JAVA_LANG_RECORD.equals(obj.getSuperclassName())) {
+            return;
+        }
+
```
Comment (KengoTODA) [inline spotbugs/src/main/java/edu/umd/cs/findbugs/detect/UncallableMethodOfAnonymousClass.java:66]: The test passes even when I revert this change. I assume that the test case is not enough to reproduce the reported problem.

```console
$ ./gradlew clean spotbugs-tests:test --tests=Issue2120Test
```

**Review by KengoTODA — CHANGES_REQUESTED** [review]
From next, please consider to make two commits in one PR: one is only for failing test cases, and next is fix with documents. Then we can use GitHub Actions to verify that test and fix are enough or not.

**Comment by baloghadamsoftware** [comment]
> From next, please consider to make two commits in one PR: one is only for failing test cases, and next is fix with documents. Then we can use GitHub Actions to verify that test and fix are enough or not.

Good idea! Thank you! I will do as you suggested.

File: spotbugs/src/main/java/edu/umd/cs/findbugs/detect/UncallableMethodOfAnonymousClass.java (hunk 1 above)
Comment (baloghadamsoftware) [reply spotbugs/src/main/java/edu/umd/cs/findbugs/detect/UncallableMethodOfAnonymousClass.java:66]: Thank you for noticing this. I forgot to add the inner classes to the test. (I always do the tests first, run them manually to see them fail, and then I do the fix.)

File: spotbugs/src/main/java/edu/umd/cs/findbugs/detect/UncallableMethodOfAnonymousClass.java [hunk 2]
```diff
@@ -59,6 +59,11 @@ public UncallableMethodOfAnonymousClass(BugReporter bugReporter) {
 
     @Override
     public void visitJavaClass(JavaClass obj) {
+        if (Values.DOTTED_JAVA_LANG_ENUM.equals(obj.getSuperclassName()) ||
+                Values.DOTTED_JAVA_LANG_RECORD.equals(obj.getSuperclassName())) {
+            return;
```
Comment (ThrawnCA) [inline spotbugs/src/main/java/edu/umd/cs/findbugs/detect/UncallableMethodOfAnonymousClass.java:64]: There could be extra methods added to an Enum that would validly benefit from the rule, though.

Comment (baloghadamsoftware) [reply spotbugs/src/main/java/edu/umd/cs/findbugs/detect/UncallableMethodOfAnonymousClass.java:64]: Hmm, you are right. I will change this to only disregard auto-generated methods.

**Review by ThrawnCA — CHANGES_REQUESTED** [review]
Changelog has a conflict.

**Review by ThrawnCA — APPROVED** [review]

**Comment by baloghadamsoftware** [comment]
[No new false positives or crashes](https://codechecker-demo.eastus.cloudapp.azure.com/java-static-analysis/reports?items-per-page=100&run=SpotBugs%20on%20MATSim-Libs%20%28user%3A%20spotbugs,%20branch%3A%20master%29&run=SpotBugs%20on%20Jenkins%20%28user%3A%20spotbugs,%20branch%3A%20master%29&run=SpotBugs%20on%20SpotBugs%20%28user%3A%20spotbugs,%20branch%3A%20master%29&run=SpotBugs%20on%20Bt%20%28user%3A%20spotbugs,%20branch%3A%20master%29&run=SpotBugs%20on%20OpenGrok%20%28user%3A%20spotbugs,%20branch%3A%20master%29&run=SpotBugs%20on%20Ttorrent%20%28user%3A%20spotbugs,%20branch%3A%20master%29&run=SpotBugs%20on%20NanoHTTPD%20%28user%3A%20spotbugs,%20branch%3A%20master%29&newcheck=SpotBugs%20on%20MATSim-Libs%20%28user%3A%20baloghadamsoftware,%20branch%3A%20Issue2120%29&newcheck=SpotBugs%20on%20Jenkins%20%28user%3A%20baloghadamsoftware,%20branch%3A%20Issue2120%29&newcheck=SpotBugs%20on%20SpotBugs%20%28user%3A%20baloghadamsoftware,%20branch%3A%20Issue2120%29&newcheck=SpotBugs%20on%20Bt%20%28user%3A%20baloghadamsoftware,%20branch%3A%20Issue2120%29&newcheck=SpotBugs%20on%20OpenGrok%20%28user%3A%20baloghadamsoftware,%20branch%3A%20Issue2120%29&newcheck=SpotBugs%20on%20Ttorrent%20%28user%3A%20baloghadamsoftware,%20branch%3A%20Issue2120%29&newcheck=SpotBugs%20on%20NanoHTTPD%20%28user%3A%20baloghadamsoftware,%20branch%3A%20Issue2120%29) on some open-source projects.

**Review by ThrawnCA — APPROVED** [review]

**Review by KengoTODA — APPROVED** [review]
Thank you!
